### ISSUE TITLE
Precache manifest inaccurate for ExtractTextPlugin output

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/workbox#1529]
(This issue was moved over from https://stackoverflow.com/questions/50836611)

**Library Affected**:
*workbox-webpack-plugin*

I'm using Angular (custom Webpack build procedure) together with the Workbox v.3.2.0 with workbox-webpack-plugin v.3.2.0 (InjectManifest plugin) for the precaching purposes and have difficulties with revisioning of the style.css file that is generated by the extract-text-webpack-plugin.

If I change the appropriate CSS files, Webpack generates updated style.css file, but its revision is not changed in the precache manifest.

As a result, I end up with the file that is not updated in the service worker cache.

Here are the related parts of the Webpack configuration (full config is too big):

```js
{
  entry: {
    'polyfills': './src/polyfills.ts',
    'main': AOT ? './src/main.browser.aot.ts' : './src/main.browser.ts',
    'style': './src/main.scss'
  },
  resolve: {
    extensions: ['.ts', '.js', 'json', '.scss', '.html'],
    modules: [helpers.root('src'), helpers.root('node_modules')]
  }, 
  module: {
    rules: [
      ...
      {
        test: /\.scss$/,
        use: ExtractTextPlugin.extract(['css-loader?sourceMap', 'postcss-loader', 'sass-loader?sourceMap']),
        exclude: [helpers.root('src', 'styles')]
      },
      ...
    ]
  },
  plugins: [
    ...
    new ExtractTextPlugin("[name].css"),
    new InjectManifest({
      swSrc: './service-worker.js',
      include: [
        /\.html$/,
        /\.js$/,
        /\.css$/,
        /\.ico$/,
        /\.json$/,
        /\.png/,
        /\.svg$/,
        /\.gif$/,
        /\.woff$/,
        /\.ttf$/,
        /\/workbox.*\/.*$/,
        /css.*\.css$/,
        /fonts.*\.(svg|eot|ttf|woff)$/,
        /i18n.*\.json$/,
        /img.*\.(svg|png|gif|jpeg|jpg)$/,
      ],
      exclude: [
        /poc\/.*$/,
        /service-worker.js$/,
        /icons-sprite/,
      ],
      importWorkboxFrom: 'local',
    }),
    ...
  ]
}
```

Please, note, that I can't use somethig like `new ExtractTextPlugin("[name].[contenthash].css")` due to some issues with caching the outdated version of the index.html referencing files from the previous releases (not existing anymore). That's why we use anti-cache URL params instead of embedding the revision in the file name.

For any other Webpack-generated file the revision is updated without any issues (even for the assets that do not have revision embedded into the filename)

Could anyone please help me out here? Thanks :)

**Comment by jeffposnick** [comment]
So what's going on here is that, currently, the webpack plugin will first check to see if a given file exists as part of a chunk, and if so, use the `renderedHash` value associated with that chunk as the effective hash in the precache manifest.

https://github.com/GoogleChrome/workbox/blob/27b8f383405ce0528227ed0c77371ec8a68e7c22/packages/workbox-webpack-plugin/src/lib/get-manifest-entries-from-compilation.js#L107-L135

If the file doesn't exist as part of a chunk, but it is an asset, then the plugin will use the asset's `source()` to find the MD5 hash, and use that as the effective hash in the precache manifest:

https://github.com/GoogleChrome/workbox/blob/27b8f383405ce0528227ed0c77371ec8a68e7c22/packages/workbox-webpack-plugin/src/lib/get-asset-hash.js#L25-L29

Looking at the config used above, the `ExtractTextPlugin` generates a `main.css` file which ends up in both the `chunks` and `assets` lists. The problem is that when the contents of the `main.css` file changes, the `renderedHash` associated with the chunk never gets updated—and the current Workbox codebase gives precedences to a chunk's `renderedHash`, using that instead of calculating each file's hash on the fly if we could avoid it. (https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/153 sounds sorta-related.)

So... I think the fix for this specific issue is to switch around that precedence, and for every asset in a compilation, generate the MD5 hash based on its `source()` on the fly, giving that precedence over a chunk's `renderedHash`.

The upside of this change would be:

- Fixing the scenario originally reported, where a chunk's `renderedHash` doesn't change but the underlying contents of a file in that chunk does change.

- Making cache expiration more fine-grained in general, since if there are multiple files in a chunk (and putting aside the special case behavior with `ExtractTextPlugin`), we are less likely to expire files that haven't changed, but ended up with a new hash by virtue of an unrelated file in that chunk changing.

The downside of this change would be:

- Calculating MD5s for all assets can add overhead in the build time, especially for larger projects.

- Maybe some other things that I'm overlooking?

**Comment by mdrideout** [comment]
In am also experiencing this issue where CSS changes do not update the service-worker on the client end. In example, I update an .scss file, webpack runs to build the new styles.css, but the client never picks up a change in the service worker (and I'm unsure if the service worker has even been changed).

This is my webpack config relevant to browser-side compilation (I have a server-side rendered react app).

```
// Webpack settings unique to browser-side script
const browserConfig = {
	entry: ['@babel/polyfill', './src/browser/index.js'],
	devtool: "source-map",
	module: {
		rules: [
			{
				test: [/\.svg$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
				loader: "file-loader",
				options: {
					name: "/media/[name].[ext]"
				}
			},
			{
				test: /\.(css|scss)$/,
				use: extractSass.extract({
					use: [
						{ loader: 'css-loader', options: { sourceMap: true } },
						{
							loader: 'postcss-loader',
							options: {
								// from create-react-app
								ident: 'postcss',
								sourceMap: true,
								plugins: () => [
									require('postcss-flexbugs-fixes'),
									autoprefixer({
										browsers: [
											'>1%',
											'last 4 versions',
											'Firefox ESR',
											'not ie < 9', // React doesn't support IE8 anyway
										],
										flexbox: 'no-2009',
									}),
								],
							},
						},
						{
							loader: 'sass-loader',
							options: {
								sourceMap: true,
								includePaths: [path.resolve(__dirname, 'node_modules')]
							}
						}
					]
				})
			},
			{
				test: /\.(js|jsx)$/,
				exclude: /node_modules/,
				use: [
					{
						loader: "babel-loader"
					},
					{
						loader: "eslint-loader",
						options: {
							quiet: true, // Ignores warnings when compiling. Warnings still show in editor. (useful for asset imports that are purely processed to public for css and not used in the script file)
						}
					}
				]
			},
			{
				test: /npm\.js$/, // BUG FIX FOR FIREBASE 5.0.2 https://github.com/firebase/firebaseui-web/issues/392
				loader: 'string-replace-loader',
				include: path.resolve('node_modules/firebaseui/dist'),
				options: {
					search: 'require(\'firebase/app\');',
					replace: 'require(\'firebase/app\').default;',
				},
			},
		]
	},
	plugins: [
		function () {
			this.plugin('watch-run', function (watching, callback) {
				console.log('\nBegin compile at ' + new Date());
				callback();
			})
		},
		new CleanWebpackPlugin(pathsToClean),
		extractSass,
		new webpack.DefinePlugin({
			'process.env.NODE_ENV': JSON.stringify('development')
		}),
		new Visualizer({
			filename: './npm-statistics-dev.html'
		}),
		new WorkboxPlugin.GenerateSW({
			swDest: 'service-worker.js',
			importsDirectory: 'wb-assets',
			runtimeCaching: [
				{
					// HOME PAGE
					urlPattern: /.+(.com|5000)\/$/,
					
					// Use stale while revalidate strategy: https://developers.google.com/web/tools/workbox/reference-docs/latest/workbox.strategies.StaleWhileRevalidate
					handler: 'staleWhileRevalidate',
					options: {
						cacheableResponse: {
							statuses: [0, 200]
						}
					}
				}
			],
			skipWaiting: true,
		})
	],
	output: {
		filename: 'bundle.js',
		path: path.resolve(__dirname, 'public')
	}
}
```

**Comment by cemulate** [comment]
I'm experiencing the same issue as @mdrideout. My config:

```
module.exports = {
    ...
    module: {
        rules: [
            ...
            {
                test: /\.scss$/,
                loader: ExtractTextPlugin.extract('css-loader!sass-loader'),
            },
            ...
        ],
    },
    plugins: [
        ...
        new workboxWebpackPlugin.GenerateSW({
            include: [ /\.html$/, /\.js$/, /\.css$/ ],
        }),
        ...
    ]
};

But when I update a `scss` file, the precache manifest hash doesn't change.

**Comment by cemulate** [comment]
Update: The issue does not occur when using [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin) instead. In addition, as of Webpack v4, [extract-text-plugin is not recommended to be used for CSS](https://github.com/webpack-contrib/extract-text-webpack-plugin#usage), with the intended replacement being the previously mentioned package.

**Comment by mdrideout** [comment]
I can confirm after upgrading to webpack 4 and using mini-css-extract-plugin, my service worker is now updating with just scss changes.

**Comment by jeffposnick** [comment]
This should be addressed by the current Workbox v5.0.0 alpha.
